### ISSUE TITLE
Multiple problems with repeated binary encoding

### ISSUE THREAD

**Issue by wvxvw** [google/protobuf#4629]
# First problem

Repeated fields aren't packed, contrary to what documentation suggests.

Proto definition:

``` protobuf
syntax = "proto3";

message TestSimple {
    repeated sint32 some_ints = 2;
}

message TestMultiple {
    repeated sfixed32 some_fixed = 3;
    repeated string some_strings = 4;
}

message Test {
    repeated TestSimple simple_repeats = 4;
    repeated TestMultiple multiple_repeats = 1;
    oneof either_or {
        TestSimple simple_oneof = 6;
        TestMultiple multiple_oneof = 5;
    }
}
```

Invoking `protoc` with this text encoded message:

```
simple_repeats: {
    some_ints: [1, -1, 2, -2, 3, -3]
}
multiple_repeats: [{
    some_fixed: [1, -1, 2, -2, 3, -3]
    some_strings: ["foo", "bar", "\x01weird\x10"]
},
{
    some_fixed: [4, -4, 5, -5, 6, -6]
}]
multiple_oneof: {
    some_fixed: [123456789, 987654321]
    some_strings: [""]
}
```

Produces this output (I've labeled it for you):

```
\x0a (field: 1/multiple_repeats of Test, wiretype: 2/lenght-delimited)
-    (length: 45)
\x1a (field: 3/some_fixed of TestMultiple, wiretype: 2/length-delimited)
\x18 (length: 24)
\x01\x00\x00\x00 1
\xff\xff\xff\xff -1
\x02\x00\x00\x00 2
\xfe\xff\xff\xff -2
\x03\x00\x00\x00 3
\xfd\xff\xff\xff -3
\x22 (field: 4/some_strings of TestMultiple, wiretype: 2/length-delimited)
\x03 (length: 3)
foo
\x22 (field: 4/some_strings of TestMultiple, wiretype: 2/length-delimited)
\x03 (length: 3)
bar
\x22 (field: 4/some_strings of TestMultiple, wiretype: 2/length-delimited)
\x07 (length: 7)
\x01weird\x10
--- end of (field: 1/multiple_repeats of Test, wiretype: 2/lenght-delimited)

\x0a (field: 1/multiple_repeats of Test, wiretype: 2/lenght-delimited)
\x1a (length: 26)
\x1a (field: 3/some_fixed of TestMultiple, wiretype: 2/length-delimited)
\x18 (length: 24)
\x04\x00\x00\x00 4
\xfc\xff\xff\xff -4
\x05\x00\x00\x00 5
\xfb\xff\xff\xff -5
\x06\x00\x00\x00 6
\xfa\xff\xff\xff -6
\x22 (field: 4/some_strings of TestMultiple, wiretype: 2/length-delimited)
\x08 (length: 8) --- should be 0.
--- end of (field: 1/multiple_repeats of Test, wiretype: 2/lenght-delimited)

\x12 (filed: 2/some_ints of TestSimple, wiretype: 2/length-delimited)
\x06 (length: 6)
\x02 -1
\x01 1
\x04 -2
\x03 2
\x06 -3
\x05 3
--- end of (filed: 2/some_ints of TestSimple, wiretype: 2/length-delimited)

* (field: 5 multiple_oneof of Test, wiretype: 2/length-delimited)
\x0c (length: 12)
\x1a (field: 3/some_fixed of TestMultiple, wiretype: 2/length-delimited)
\x08 (length: 8)
\x15\xcd[\x07 123456789
\xb1h\xde: 987654321
--- end of (field: 5 multiple_oneof of Test, wiretype: 2/length-delimited)

\x22 (field: 4/some_strings of TestMultiple, wiretype: 2/length-delimited)
\x00 (length: 0)
```

# Second problem

See my comment above on incorrectly encoding the length of the string that wasn't specified on input. Its length was set to 8, but, probably the parser ignores this value because it already knows the overall length of the message.

----

Some more general comments:

Encoding repeated fields the way you do in the example above makes parsing messages from stream impossible / meaningless because the parser needs to wait until the entire message is received and processed in order to parse any of it.  This also impairs the parser in that it absolutely must follow your implementation instead of specification. The ability to concatenate message payloads and then parse the result is completely useless (I believe it happened to you by chance that it worked this way, but then you decided to document this bizarre behavior.)  This also prevents alternative implementations from creating more efficient parsers because, except for this part one could assume that a single type specification in parser would map to unique binary representation.  But since this assumption is broken by the fact that you can encode the same value by either repeating field entry multiple times or by writing all field values together, you can no longer use the proto definition as a guide for what to expect from the binary payload.

Bottom line: multiple ways to encode repeated fields is a very bad idea.  Please don't do that.

**Comment by acozzette** [comment]
I believe everything is actually working as intended here. When you see a field number 4 with a length of 8, that's not the `some_strings` field of `TestMultiple` but instead it's the `simple_repeats` field of `Test`. The empty string doesn't appear until the end, since the serialized output is sorted by field number.

I think your other objection is that once you've read a field, you can't know for sure if it's the real value, because it could be appended or overridden later on. That's true, but at least in practice it would be somewhat rare for this to happen, so you can optimize for the common case of each field being set in one place.

You are right that there is not a unique binary representation for a given message. This is especially for true if you use map fields, because maps are logically unordered and so the order of the serialized entries is arbitrary. In some languages we have APIs for deterministic serialization since some use cases require that, but we only guarantee that the serialization will be deterministic for a given binary (it might still change over time with new protobuf versions).

**Comment by wvxvw** [comment]
Your analysis re' `some_strings` vs `simple_repeats` is wrong. Sorry. You can absolutely see that the contents is the strings given on the stdout, and not varints, as would have been, should this actually encode `simple_repeats`.

Re' non-uniqueness of representation: lack of order is acceptable.  If you go further into theoretical properties of the information being generated, you can see that allowing permutations of information is a much weaker kind of inconsistency than allowing two entirely different encodings of atomic information.  This is similar to how we (humans) are normally OK with sums of integers written as `1+2` or `2+1` (and meaning the same thing), but will have problems accepting systems with logical rules like `2=1` and `1=1` (both simultaneously true).

----

So, my other reading of your response is that: you are actually never packing length-delimited repeated fields, so it is safe for alternative implementation to assume that this will be the case for a while (and probably until next major version)?  I think that this is inefficient and creates some useless, and even harmful properties of Protobuf encoding, but if that's how you like it, who am I to tell you otherwise.

**Comment by wvxvw** [comment]
This is not the part I'm writing about... the so-called scalar types are the only types that work according to the description in documentation, the length-delimited types are the problem.

**Comment by acozzette** [comment]
The thing is that only repeated fields of primitive numeric types can be packed, so for example, the repeated strings in your example can't be packed.

**Comment by acozzette** [comment]
I agree that in principle one could pack all types of repeated fields, but this is not something allowed by the current wire format. As the [documentation](https://developers.google.com/protocol-buffers/docs/encoding#packed) says: 'Only repeated fields of primitive numeric types (types which use the varint, 32-bit, or 64-bit wire types) can be declared "packed".' We would be pretty reluctant to make length-delimited fields packable, because existing parsers would be unable to parse messages in the new format. Requiring repeated fields to appear in one continuous field would also break existing parsers.

Allowing multiple mentions of the same repeated field is also not entirely useless. Let's say you want to merge two protos together, and each one happens to have a length-delimited unknown field of the same field number. Those unknown fields might be packed repeated fields, but we have no way of knowing that (they could also be strings or messages for example). That means that when we reserialize the resulting merged proto, it's impossible to pack the unknown fields into your suggested canonical format. The only safe thing to do is serialize them as separate tag-value pairs.

**Comment by wvxvw** [comment]
> Let's say you want to merge two protos together

This never happens in practice.  All people I talked to so far about Protobuf are horrified even when they hear that it is possible, in principle, to use two different definitions for the same binary payload: they usually start bursting into flames when they hear such thing.  Merging messages for them makes absolutely no sense.  Most people think about any message format, not necessary Protobuf, as if it was encoding one conceptual message into one binary representation.  Even if there is a possibility to do it differently, nobody would use that.

From a perspective of a user of Protobuf, who is normally a Java / C++ programmer, they just need to put a definition of messages into their source code to make sure that classes on one end of the channel translate into same classes on the other end.  Merging messages in this context would be analogous to be able to merge objects of arbitrary classes, say, `Animal` and `Transaction`.  This just makes one's mind explode :)

**Comment by acozzette** [comment]
People do merge protos in practice: you can search our GitHub issues for examples.

Protocol buffers are designed to make it easy to continuously evolve your schema over time without breaking compatibility. This makes it inevitable that the parser will be somewhat generous in what it accepts, because it never assumes it has the most up-to-date version of the schema.

**Comment by xfxyjwf** [comment]
I think you folks are talking about different things. @acozzette is talking about merging protobuf data (see Message::MergeFrom: https://developers.google.com/protocol-buffers/docs/encoding#optional), whereas @wvxvw misunderstood it to be merging two different proto types.

**Comment by wvxvw** [comment]
@xfxyjwf Well, there is a misunderstanding, however, I think it's different: proto types exist on a much higher level of the parser then where merging happens.  So, there's nothing to stop you from merging chunks of data that would have mapped to different C++ types before you interpret them as such.

What I'm saying is that this ability is not just not very useful, is that it is also dangerous, and it's danger haven't been discovered because, mostly people using Protobuf don't even realize it is possible to use it like that.

I am convinced that dropping this feature in the next version in favor of packed length-delimited fields would've made everyone happier: you get more compact wire format, you get an ability to generate pull/push parsers, you have a more uniform binary parser and all you have to pay for this is an ability to merge messages.  Seems like a huge win to me.

**Comment by acozzette** [comment]
@wvxvw I am afraid we just can't make breaking changes like that--that would result in all newly serialized protos being unparseable by existing parsers.

**Comment by xfxyjwf** [comment]
Closing as there is nothing we can do here.

@wvxvw If you are creating a new wire format for protobuf, I would suggest make it completely incompatible with the current proto wire format rather than partially compatible. For example, you probably want to use different wire types.
